src/AutoMapper/TypePair.cs

```diff
@@ -185,6 +180,13 @@ public int Compare(Type x, Type y)
                 return 0;
             }
         }
+    }
 
+    public static class HashCodeCombiner
+    {
+        public static int Combine<T1, T2>(T1 obj1, T2 obj2) =>
+            CombineCodes(obj1.GetHashCode(), obj2.GetHashCode());
+
+        public static int CombineCodes(int h1, int h2) => ((h1 << 5) + h1) ^ h2;
```

Comment:
Actually no, because sometimes you'll compute it when it's not needed. Try the simple benchmark with your changes and compare with my version.